### ISSUE TITLE
Problems scrubbing through frame

### ISSUE THREAD

**Issue by bhackett1024** [RecordReplay/devtools#1815]
See comment in this recording: https://replay.io/view?id=36dc99a2-9bf0-4314-b743-0ccda8c549c8

A frame whose behavior I was interested in is pretty messed up when trying to scrub or step through it, with paused lines and control flow that makes no sense.  This happens both with the original source and when switching to view the generated source.

**Comment by bhackett1024** [comment]
This might be a separate issue, but there are also problems setting logpoints in the frame with the messed up scrubbing, see the comment in the recording.

**Comment by jasonLaster** [comment]
@bhackett1024 what priority do you think this should be?

**Comment by bhackett1024** [comment]
It would be nice to get to this soon but it's non-trivial to investigate and we have a lot of other stuff on our plate atm.

**Comment by jasonLaster** [comment]
I'll bump it to next week and we can discuss during planning

**Comment by hbenl** [comment]
I found a way to consistently reproduce this:
Open a recording of Replay (e.g. https://replay.io/view?id=4babe7ce-f6c1-4104-a4b3-a64065051820), open `devtools/client/webconsole/reducers/messages.js`, set a logpoint at line 917 and jump to any of the logpoint hits.
* The scope shows the variable `timeStampSort`, which should be a boolean, to be a redux action:

![Screenshot_2021-03-04_18-27-46](https://user-images.githubusercontent.com/16060807/110004784-18946e80-7d18-11eb-9306-5f3c047bbc26.png)

* Scrubbing through that frame makes weird jumps through the code.
* Go up one stackframe and you'll see that this was actually a call to the messages reducer:

![Screenshot_2021-03-04_18-28-35](https://user-images.githubusercontent.com/16060807/110005107-6b6e2600-7d18-11eb-80c0-9afe56c3c52a.png)

* But if you set a logpoint in the messages reducer, only some of the calls to it are shown:

![Screenshot_2021-03-04_18-42-17](https://user-images.githubusercontent.com/16060807/110005972-5d6cd500-7d19-11eb-9fc6-214d00fe893a.png)

**Comment by jasonLaster** [comment]
@hbenl could you record this and leave comments?

**Comment by hbenl** [comment]
I have added a comment to https://replay.io/view?id=4babe7ce-f6c1-4104-a4b3-a64065051820 that takes you straight to one of the broken frames.

**Comment by bhackett1024** [comment]
> @bhackett1024 I think this is a backend issue, what do you think?

Probably, but it's hard to be sure.  Can you find the message which the backend sent which has incorrect information?  That will help in identifying where things went wrong in the backend.

**Comment by hbenl** [comment]
So here's a recording of me scrubbing through the broken frame in the previous recording: https://replay.io/view?id=b47e0b25-6bab-4c54-8715-47b201807f10
I have added comments showing where the frameSteps are received and what scrubbing through the frame using both original and pretty-printed sources looks like. Note that the steps in the original source don't fully match the steps in the pretty-printed source.
Go to the first comment and extract the line numbers for the frameSteps in the original source using `steps.map(s => s.frame[2].line)` or the pretty-printed source using `steps.map(s => s.frame[1].line)`. Those line numbers correspond to the scrubbing positions in both. So it looks to me that the devtools' visualization of the data from the backend is correct and it's this data that is incorrect (perhaps due to a sourcemapping issue?).

**Comment by hbenl** [comment]
I've noticed that this only happens in recordings of https://replay.io, but not recordings of http://localhost:8080. Perhaps because the generated sourcemaps are different for dev and prod?

**Comment by bhackett1024** [comment]
Yeah, this is probably an issue related to the production source maps.

**Comment by loganfsmyth** [comment]
This is several things happening at once, all for the same reason.

1. The `messages.js` positions are broken because of https://github.com/terser/terser/issues/960 We load `whatwg-url` to parse URLs, and that contains a lookup table that has some non-BMP characters on it, and it just so happens that when we minify our output (because we use `webpack --mode=production`, the `message.js` file's code is placed on the same output line as `whatwg-url`.
2. Jumping to `main.js` like is done in the second half of https://replay.io/view?id=b47e0b25-6bab-4c54-8715-47b201807f10 is broken because our editor itself also expects line/column offsets to be code-unit offsets, not code point offsets, so the positions we get back for that are correct, but then the editor takes you to the wrong place.

**Comment by loganfsmyth** [comment]
Another thing we need to do is verify how our pretty-printer handles non-BMP characters.

**Comment by loganfsmyth** [comment]
Alright as I suspected, pretty-printed sourcemaps also had the same issue. With https://github.com/RecordReplay/backend/pull/1543 you can at least scrub through the pretty-printed generated code properly.

**Comment by jasonLaster** [comment]
@loganfsmyth this looks fixed with #1543. Is that correct?

**Comment by loganfsmyth** [comment]
Yeah, I had left this open at the time because I had hoped that we'd get a response on https://github.com/terser/terser/issues/960 but we haven't. Your call if you want to leave this open for that or close.
